File: backends/fastapi/app/services/admin_service.py
"""
管理员认证服务
提供一个足够简单、便于教学的 JWT 登录流程。
"""
from datetime import datetime, timedelta, timezone

import jwt

from app.config import ADMIN_PASSWORD, JWT_SECRET, JWT_EXPIRATION_HOURS


class UnauthorizedException(Exception):
    """认证失败"""
    def __init__(self, message: str = "认证失败"):
        super().__init__(message)


def login(password: str) -> str | None:
    """
    管理员登录。
    演示项目只保留一个固定管理员身份，因此密码通过后直接签发 token。
    """
    if password != ADMIN_PASSWORD:
        return None

    now = datetime.now(timezone.utc)
    payload = {
        "sub": "admin",
        "iat": now,
        "exp": now + timedelta(hours=JWT_EXPIRATION_HOURS),
    }
    return jwt.encode(payload, JWT_SECRET, algorithm="HS256")


def validate_token(token: str) -> bool:
    """验证 JWT 令牌是否可用。"""
    try:
        jwt.decode(token, JWT_SECRET, algorithms=["HS256"])
        return True
    except (jwt.InvalidTokenError, Exception):
        return False

File: backends/gin/service/admin_service.go
// Package service 管理员认证服务
// JWT 令牌的生成与验证
package service

import (
	"time"

	"github.com/alexhe/hellotime-gin/config"
	"github.com/golang-jwt/jwt/v5"
)

// Login 管理员登录
// 密码正确返回 JWT token，否则返回空字符串
func Login(password string) string {
	if password != config.AdminPassword {
		return ""
	}

	now := time.Now().UTC()
	claims := jwt.MapClaims{
		"sub": "admin",
		"iat": now.Unix(),
		"exp": now.Add(time.Duration(config.JWTExpirationHours) * time.Hour).Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString([]byte(config.JWTSecret))
	if err != nil {
		return ""
	}
	return tokenString
}

// ValidateToken 验证 JWT 令牌
func ValidateToken(tokenString string) bool {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, jwt.ErrSignatureInvalid
		}
		return []byte(config.JWTSecret), nil
	})
	if err != nil {
		return false
	}
	return token.Valid
}
